types/src/addressable_entity.rs

```diff
@@ -658,6 +676,761 @@ impl KeyValueJsonSchema for EntryPointLabels {
+            _ => unreachable!(),
+        }
+    }
+}
+
+// pub struct QueryAddressableEntity {
+//     addressable_entity: AddressableEntity,
+//     named_keys: NamedKeys
+// }
+
+/// The address for an AddressableEntity which contains the 32 bytes and tagging information.
+#[derive(PartialOrd, Ord, PartialEq, Eq, Hash, Clone, Copy, Serialize, Deserialize)]
+#[cfg_attr(feature = "datasize", derive(DataSize))]
+#[cfg_attr(feature = "json-schema", derive(JsonSchema))]
+pub enum EntityAddr {
+    /// The address for a system entity account or contract.
+    System(HashAddr),
+    /// The address of an entity that corresponds to an Account.
+    Account(HashAddr),
+    /// The address of an entity that corresponds to a Userland smart contract.
+    SmartContract(HashAddr),
+}
+
+impl EntityAddr {
+    /// The length in bytes of a [`EntityAddr`].
+    pub const ENTITY_ADDR_LENGTH: usize = U8_SERIALIZED_LENGTH + KEY_HASH_LENGTH;
+
+    /// Constructs a new [`EntityAddr`] for a system entity.
+    pub const fn new_system_entity_addr(hash_addr: [u8; KEY_HASH_LENGTH]) -> Self {
+        Self::System(hash_addr)
+    }
+
+    /// Constructs a new [`EntityAddr`] for an Account entity.
+    pub const fn new_account_entity_addr(hash_addr: [u8; KEY_HASH_LENGTH]) -> Self {
+        Self::Account(hash_addr)
+    }
+
+    /// Constructs a new [`EntityAddr`] for a smart contract.
+    pub const fn new_contract_entity_addr(hash_addr: [u8; KEY_HASH_LENGTH]) -> Self {
+        Self::SmartContract(hash_addr)
+    }
+
+    /// Constructs a new [`EntityAddr`] based on the supplied tag.
+    pub fn new_with_tag(entity_kind: EntityKind, hash_addr: [u8; KEY_HASH_LENGTH]) -> Self {
+        match entity_kind {
+            EntityKind::System(_) => Self::new_system_entity_addr(hash_addr),
+            EntityKind::Account(_) => Self::new_account_entity_addr(hash_addr),
+            EntityKind::SmartContract => Self::new_contract_entity_addr(hash_addr),
+        }
+    }
+
+    /// Returns the tag of the [`EntityAddr`].
+    pub fn tag(&self) -> EntityKindTag {
+        match self {
+            EntityAddr::System(_) => EntityKindTag::System,
+            EntityAddr::Account(_) => EntityKindTag::Account,
+            EntityAddr::SmartContract(_) => EntityKindTag::SmartContract,
+        }
+    }
+
+    /// Returns the 32 bytes of the [`EntityAddr`].
+    pub fn value(&self) -> HashAddr {
+        match self {
+            EntityAddr::System(hash_addr)
+            | EntityAddr::Account(hash_addr)
+            | EntityAddr::SmartContract(hash_addr) => *hash_addr,
+        }
+    }
+
+    /// Returns the formatted String representation of the [`EntityAddr`].
+    pub fn to_formatted_string(&self) -> String {
+        format!("{}", self)
+    }
+
+    /// Constructs an [`EntityAddr`] from a formatted String.
+    pub fn from_formatted_str(input: &str) -> Result<Self, FromStrError> {
+        if let Some(entity) = input.strip_prefix(ENTITY_PREFIX) {
```

Comment:
Added the test, which is passing

Reply:
Ah, it works because both `ADDRESSABLE_ENTITY_STRING_PREFIX` and `ENTITY_PREFIX` are the same string (`addressable-entity-`). It would be good to use only one for consistency.